Doc gatekeeper operator 2.2

dockerymick opened the pull request:
https://github.com/open-cluster-management/backlog/issues/7433
https://github.com/open-cluster-management/backlog/issues/6274

dockerymick committed:
adding doc for gatekeeper operator

dockerymick committed:
small update

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -7,6 +7,13 @@ Learn to create, apply, view, and update your gatekeeper policies.
+* <<creating-a-gatekeeper-policy,Creating a gatekeeper policy>>
+* <<creating-a-gatekeeper-policy-for-admission,Creating a gatekeeper policy for admission>>
+* <<creating-a-gatekeeper-policy-for-audit,Creating a gatekeeper policy for audit>>
+* <<gatekeeper-operator,Creating a gatekeeper operator>>
```
this should appear first as it installs the gatekeeper. We can remove `Prerequisite` as this is what this policy does

ycao56 replied:
I would rephrase the title to `Install gatekeeper using gatekeeper operator policy`

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -163,5 +170,22 @@ spec:
+[#gatekeeper-operator]
+=== Creating a gatekeeper operator
+
+Use the governance framework to install the gatekeeper operator (`policy-gatekeeperoperator`). Gatekeeper operator is available in the link: https://operatorhub.io/[OperatorHub.io community].
```
RHACM provides gatekeeper operator policy for user to install gatekeeper on managed cluster. 
```
Gatekeeper operator is available in the link: https://operatorhub.io/[OperatorHub.io community].
```
The version we ship is Red Hat supported version not community version so we shouldn't point it to operatorhub. It will be available in ocp catalog.

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -163,5 +170,22 @@ spec:
+
+During the install, the operator group and subscription pull the Gatekeeper operator to install it on to your hub cluster.
+
+Gatekeeper operator policy is monitored by the Red Hat Advanced Cluster Management configuration policy controller, where `enforce` remediation action is supported. Gatekeeper policies are created automatically by the controller when set to `enforce`.
```
`Gatekeeper operator policy` not `Gatekeeper policies`

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -163,5 +170,22 @@ spec:
+. Log in to your cluster.
+. From the navigation menu, click Govern risk
+. Click *Create policy*.
+. As you complete the form, select *GatekeeperOperator* from the _Specifications_ field. The parameter values for your policy are automatically populated.
```
the policy needs to be enforced in order to install gatekeeper

ycao56 commented:
We need to document more about `gatekeeper operator` e.g. how it works and all the parameters that we can configure. @cadawson was on the gatekeeper operator sync meeting call. @dockerymick can you sync with him to see what's our plan to document gatekeeper operator?

dockerymick committed:
Update create_gatekeeper.adoc

Changes made after Yu's suggestion

dockerymick committed:
Update gatekeeper_policy.adoc

updated the title name for the file referenced at the end of this doc

dockerymick committed:
Update create_gatekeeper.adoc

Added a reference to the OCP documentation and added brief explanation of Gatekeeper functionality

dockerymick committed:
Update create_gatekeeper.adoc

dockerymick commented:
> We need to document more about `gatekeeper operator` e.g. how it works and all the parameters that we can configure. @cadawson was on the gatekeeper operator sync meeting call. @dockerymick can you sync with him to see what's our plan to document gatekeeper operator?

@ycao How do you want to display the parameters? Usually, we create a YAML table with parameter definitions. Then we can reference the user to the Gatekeeper operator policy examples.

dockerymick committed:
Update create_gatekeeper.adoc

Added more info about the gatekeeper operator; I wasn't sure on if we should have the parameter table or only point to the repo

willkutler commented on security/create_gatekeeper.adoc:
```diff
@@ -1,11 +1,175 @@
+
+Use the governance framework to install the Gatekeeper operator (`policy-gatekeeperoperator`). Gatekeeper operator is available in the OpenShift Container Platform catalog. See _Adding Operators to a cluster_ in the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.6/html/operators/administrator-tasks#olm-adding-operators-to-a-cluster[OpenShift Container Platform d ...
+
+During the install, the operator group and subscription pull the Gatekeeper operator to install it on to your hub cluster. Also consider that default values can be generated by the operator. See the default YAML file for the Gatekeeper operator and review the parameter descriptions: 
```
small change but you might want to specify that this is the YAML for the Gatekeeper custom resource, rather than the operator. Also, it could cause some confusion that this YAML contains a lot of fields that aren't in the custom resource that gets created by the `policy-gatekeeperoperator` policy in the UI - we should probably clarify that a lot of these fields are optional, since just using
```
              apiVersion: operator.gatekeeper.sh/v1alpha1
              kind: Gatekeeper
              metadata:
                name: gatekeeper
                namespace: gatekeeper-system
              spec:
                audit:
                  logLevel: INFO
                  replicas: 1
                image:
                  image: 'docker.io/openpolicyagent/gatekeeper:v3.2.2'
                validatingWebhook: Enabled
                webhook:
                  emitAdmissionEvents: Enabled
                  logLevel: INFO
                  replicas: 2
```
will still install gatekeeper.

dockerymick replied:
Thanks, i've improved the statement

willkutler commented on security/create_gatekeeper.adoc:
```diff
@@ -120,16 +283,14 @@ spec:
 
-+
 ----
 apiVersion: policy.open-cluster-management.io/v1
```
Any reason the audit and admission scenarios are split into two policies? Typically for sample gatekeeper policies we've just added the audit configurationpolicy as another policy-template to go along with the admissions one. If you'd like to split them up, both policies would need to contain the configurationPolicy:
```
        apiVersion: policy.open-cluster-management.io/v1
        kind: ConfigurationPolicy
        metadata:
          name: policy-gatekeeper-k8srequiredlabels
        spec:
          remediationAction: enforce # will be overridden by remediationAction in parent policy
          severity: low
          object-templates:
            - complianceType: musthave
              objectDefinition:
                apiVersion: templates.gatekeeper.sh/v1beta1
                kind: ConstraintTemplate
                metadata:
                  name: k8srequiredlabels
                spec:
                  crd:
                    spec:
                      names:
                        kind: K8sRequiredLabels
                      validation:
                        # Schema for the `parameters` field
                        openAPIV3Schema:
                          properties:
                            labels:
                              type: array
                              items: string
                  targets:
                    - target: admission.k8s.gatekeeper.sh
                      rego: |
                        package k8srequiredlabels
                        violation[{"msg": msg, "details": {"missing_labels": missing}}] {
                          provided := {label | input.review.object.metadata.labels[label]}
                          required := {label | label := input.parameters.labels[_]}
                          missing := required - provided
                          count(missing) > 0
                          msg := sprintf("you must provide labels: %v", [missing])
                        }
            - complianceType: musthave
              objectDefinition:
                apiVersion: constraints.gatekeeper.sh/v1beta1
                kind: K8sRequiredLabels
                metadata:
                  name: ns-must-have-gk
                spec:
                  match:
                    kinds:
                      - apiGroups: [""]
                        kinds: ["Namespace"]
                  parameters:
                    labels: ["gatekeeper"]
```
since that policy is the one that contains the Rego rule for gatekeeper to use - otherwise, it won't create the `K8sRequiredLabels` object that this policy is checking for violations.

dockerymick replied:
Good idea, we wanted to highlight that there were more than one resource created in the gatekeeper policy. After taking a look again, we were assuming that the user knew that `kind: ConfigurationPolicy` was required. That isn't fair to the user, so I will add your suggestion

dockerymick replied:
Re-reviewed the file and confirmed with @willkutler that the ConfigurationPolicy is present in both examples.

dockerymick committed:
Update create_gatekeeper.adoc

changes after Will's review

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -1,11 +1,175 @@
+= Install Gatekeeper using the gatekeeper operator policy
 
-Learn to create, apply, view, and update your gatekeeper policies.
+Use the Gatekeeper operator to install the community version of Gatekeeper on a managed cluster. you can create and configure a Gatekeeper operator policy to monitor customer resources (CR) that are created in the same namespace it is deployed.
```
The policy will be updated to RHACM supported one once we published it. So it should not say community version.

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -1,11 +1,175 @@
 [#gatekeeper-policy-integration]
-= Gatekeeper policy integration
+= Install Gatekeeper using the gatekeeper operator policy
```
I think the title should be changed. Because what we are doing here is:
1.  Install gatekeeper operator on managed cluster
2. Gatekeeper operator then installs gatekeeper
3. create a gatekeeper policy
4.  report violations for audit and admission
So it is broader scope than `Install Gatekeeper using the gatekeeper operator policy`

dockerymick replied:
What do you think about `Integrating Gatekeeper`

ycao56 replied:
sounds good

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -1,11 +1,175 @@
+
+Use the governance framework to install the Gatekeeper operator (`policy-gatekeeperoperator`). Gatekeeper operator is available in the OpenShift Container Platform catalog. See _Adding Operators to a cluster_ in the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.6/html/operators/administrator-tasks#olm-adding-operators-to-a-cluster[OpenShift Container Platform d ...
+
+During the install, the operator group and subscription pull the Gatekeeper operator to install it on to your hub cluster. Also consider that default values can be generated by the operator. 
```
During the install, the operator group and subscription pull the Gatekeeper operator to install it on to your hub cluster. **Then gatekeeper operator will install and configure gatekeeper based on Gatekeeper CR.**

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -1,11 +1,175 @@
+[#gatekeeper-operator-policy]
+=== Creating a gatekeeper operator
+
+Use the governance framework to install the Gatekeeper operator (`policy-gatekeeperoperator`). Gatekeeper operator is available in the OpenShift Container Platform catalog. See _Adding Operators to a cluster_ in the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.6/html/operators/administrator-tasks#olm-adding-operators-to-a-cluster[OpenShift Container Platform d ...
```
I didn't find gatekeeper operator policy. Are we going to provide a link to it?

dockerymick replied:
Thanks for pointing that out, i have added the link to the policy in the community folder. Is that the correct link?

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -19,13 +183,12 @@ Use the {product-title-short} configuration policy to create a gatekeeper policy
-. Create a YAML file for your gatekeeper policy.
-Run the following command:
-+
+Create a YAML file for your gatekeeper policy. Run the following command:
```
the content in `Creating a gatekeeper policy` is not correct. 
The gatekeeper policy itself is missing. I am talking about https://github.com/open-cluster-management/policy-collection/blob/master/community/CM-Configuration-Management/policy-gatekeeper-sample.yaml#L12-L65 . This is how you can create a gatekeeper policy.

ycao56 commented on security/create_gatekeeper.adoc:
```diff
@@ -1,11 +1,175 @@
+. From the navigation menu, click *Govern risk*.
+. Create a policy by selecting *Create policy*.
+. As you complete the form, select *GatekeeperOperator* from the _Specifications_ field. The parameter values for your policy are automatically populated and the policy is set to enforce to install Gatekeeper.
+
```
the content in `Creating a gatekeeper policy` is not correct. 
The gatekeeper policy itself is missing. I am talking about https://github.com/open-cluster-management/policy-collection/blob/master/community/CM-Configuration-Management/policy-gatekeeper-sample.yaml#L12-L65 . This is how you can create a gatekeeper policy.

dockerymick committed:
Update create_gatekeeper.adoc

dockerymick committed:
Update create_gatekeeper.adoc

dockerymick committed:
Update gatekeeper_policy.adoc

updating title for the file

dockerymick committed:
Update create_gatekeeper.adoc

updated Creating a gatekeeper policy section

dockerymick committed:
Update gatekeeper_policy.adoc

Removed prerequisites bc gatekeeper operator policy installs gatekeeper for the user

other clean up

dockerymick committed:
Update gatekeeper_policy.adoc

added content from create_gatekeeper to improve the flow

dockerymick committed:
Update create_gatekeeper.adoc

(link:https://github.com/open-cluster-management/policy-collection/blob/master/community/CM-Configuration-Management/policy-gatekeeper-operator.yaml[`policy-gatekeeper-operator`])

dockerymick committed:
Merge branch '2.2_stage' into 6274

dockerymick commented:
@ycao56 I have made more updates to both files. I moved the information related to the content (i.e the gatekeeper table w/optional parameters). When you have an opportunity, may you re-review the PR

ycao56 commented:
LGTM. The only thing is the gatekeeper yaml in policy-collection repo will change. Currently the code snippet is referenced by lines. If the policy change in policy-colleciton repo, we will have to come back and update the doc

ycao56 commented on security/gatekeeper_policy.adoc:
```diff
@@ -1,7 +1,7 @@
 = Gatekeeper policy
 
-Gatekeeper is a validating webhook that enforces CustomResourceDefinition (CRD) based policies that are run with the Open Policy Agent (OPA). You can install Gatekeeper to integrate a gatekeeper policy with {product-title}. Gatekeeper policy can be used to evaluate Kubernetes resource compliance. You can leverage an OPA as the policy engine, and use Rego as the policy language.
+Gatekeeper is a validating webhook that enforces CustomResourceDefinition (CRD) based policies that are run with the Open Policy Agent (OPA). You can install gatekeeper on your cluster by using the gatekeeper operator policy. Gatekeeper policy can be used to evaluate Kubernetes resource compliance. You can leverage an OPA as the policy engine, and use Rego as the policy language.
```
I suggest we reorder the structure a little bit to better explain
1. Install gatekeeper using gatekeeper operator policy
2. Apply gatekeeper constraints and constraints template and monitor violation for audit and admission scenario

dockerymick committed:
Update gatekeeper_policy.adoc

updates

dockerymick committed:
Update create_gatekeeper.adoc

dockerymick committed:
Update whats_new.adoc

added explanation onto what's new

dockerymick committed:
Update gatekeeper_policy.adoc

Updates after meeting with Yu

dockerymick committed:
Update create_gatekeeper.adoc

dockerymick committed:
Update third_party_policy.adoc

Updating the title for reference link

dockerymick committed:
Update whats_new.adoc

dockerymick committed:
Update gatekeeper_policy.adoc

some changes

dockerymick committed:
Update create_gatekeeper.adoc

dockerymick committed:
Update whats_new.adoc

Updating name for doc

dockerymick committed:
Update create_gatekeeper.adoc

some changes

dockerymick committed:
Update gatekeeper_policy.adoc

updated the indentation for the YAML examples

dockerymick commented:
@ycao56 may you re-review the PR, i have addressed the comments and updates

ycao56 commented on security/gatekeeper_policy.adoc:
```diff
@@ -1,366 +1,119 @@
++
+[source,yaml]
+----
+- objectDefinition:
```
update the code snippet with https://github.com/open-cluster-management/policy-collection/blob/master/community/CM-Configuration-Management/policy-gatekeeper-sample.yaml#L67-L82

ycao56 replied:
remove this line

ycao56 commented on security/gatekeeper_policy.adoc:
```diff
@@ -1,366 +1,119 @@
-  clusterSelector:
-    matchExpressions:
-      []  # selects all clusters if not specified
+- objectDefinition:
```
remove this line

ycao56 commented:
@dockerymick looks good. just need to remove two lines in code snippets. approved!

dockerymick committed:
Update gatekeeper_policy.adoc

removed `- objectDefinition:` from the the examples

dockerymick committed:
Update whats_new.adoc

updated the title for reference link

dockerymick committed:
Update create_gatekeeper.adoc

updating anchor

dockerymick committed:
Update create_gatekeeper.adoc

small update

dockerymick committed:
changes after running the spellcheck tool
